### PR TITLE
[move][stdlib] Implement mem::swap native move call

### PR DESCRIPTION
**Pull request by igor-aptos**
## Description
If we have a field that contains non-copyable type, it is impossible to change it, and get the old value back. 

Adding two methods:
* native mem::swap, that implements swap of contents of two mutable references
* mem::replace, as a simple wrapper based on mem::swap 

## How Has This Been Tested?
provided unit tests

## Type of Change
- [x] New feature

## Which Components or Systems Does This Change Impact?
- [x] Move/Aptos Virtual Machine

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by lightmark** [comment]
What's this for?

**Comment by igor-aptos** [comment]
This is needed to be able to modify fields, when they hold a non-copyable non-dropable type.

For example, you cannot do this today:

```
struct A<T> {
   field: T,
}

public fun swap<T>(self: &mut A<T>, value: T): T {
   let old_value = self.field;  <-- doesn't compile, expects copy
   self.field = value;          <-- doesn't compile, expects drop
   old_value
}
```

https://aptos-org.slack.com/archives/C05V6JEKZQ9/p1727393314700309

**Review by wrwg — APPROVED** [review]

File: aptos-move/framework/move-stdlib/src/natives/mem.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    if args.len() != 2 {
+        return Err(SafeNativeError::InvariantViolation(PartialVMError::new(
+            StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+        )));
+    }
+
+    let cost = MEM_SWAP_BASE
+        + MEM_SWAP_PER_ABS_VAL_UNIT
+            * (context.abs_val_size(&args[0]) + context.abs_val_size(&args[1]));
```
Comment (vgao1996) [inline aptos-move/framework/move-stdlib/src/natives/mem.rs:48]: You need to use `abstract_value_size_dereferenced` as it is the sizes of the referenced values that matter, not the size of the references themselves.
https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/aptos-gas-schedule/src/gas_schedule/misc.rs#L293

You may need to add a helper to expose it through the native context.

Comment (runtian-zhou) [reply aptos-move/framework/move-stdlib/src/natives/mem.rs:48]: Why dereferenced value? We are not allocating new memory space?

Comment (vgao1996) [reply aptos-move/framework/move-stdlib/src/natives/mem.rs:48]: Because the current implementation effective makes deep clones of the two values being swapped. 

Anyway, this discussion may no longer be relevant as I've come up with a more efficient constant-time implementation. Will share the PR very very soon.

File: aptos-move/framework/move-stdlib/tests/mem_tests.move [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#[test_only]
+module std::mem_tests {
```
Comment (vgao1996) [inline aptos-move/framework/move-stdlib/tests/mem_tests.move:2]: The tests here are inadequate considering how important and dangerous the swap feature is.

We need to have full test coverage for all these swaps:
```
(container, container)

(indexed ref - generic, indexed ref - generic)
(indexed ref - specialized, indexed ref - specialized)
(indexed ref - generic, indexed ref - specialized)
(indexed ref - specialized, indexed ref - generic)
```
I would also encourage us add (rust-level) unit tests in `move-vm/types/src/values/value_tests`.

File: execution/executor/tests/internal_indexer_test.rs [hunk 3]
```diff
@@ -192,6 +193,208 @@ fn test_db_indexer_data() {
...
-    let address_one_kv_res: Vec<_> = address_one_kv_iter.collect();
-    assert_eq!(address_one_kv_res.len(), 152);
+    let address_one_kv_res = address_one_kv_iter.collect::<Result<Vec<_>, _>>().unwrap();
+
+    let (code, resources): (Vec<_>, Vec<_>) = address_one_kv_res
+        .into_iter()
+        .map(|(s, _)| s)
+        .partition(|s| s.is_aptos_code());
+
+    let expected_code = vec![
```
Comment (vgao1996) [inline execution/executor/tests/internal_indexer_test.rs:203]: Are these changes related? Not familiar with the code here and don't know what it is trying to achieve.

**Review by vgao1996 — DISMISSED** [review]
Okay, I ended up just implementing the more efficient swap as I originally promised:
- https://github.com/aptos-labs/aptos-core/pull/14921

Please take a careful look and help me check if the implementation is correct. I also just realized that `swap` is a really powerful feature, and we must take utmost caution to ensure its implementation is correct, or very bad things could happen. For this reason, I'd like to request at least one additional review from @runtian-zhou @ziaptos or @georgemitenkov + a third party audit.

**Comment by vgao1996** [comment]
@wrwg I'm also wondering if we should just make this a bytecode instruction. Can probably do it later though.

File: execution/executor/tests/internal_indexer_test.rs (hunk 3 above)
Comment (igor-aptos) [reply execution/executor/tests/internal_indexer_test.rs:203]: there is a check that all resources published in genesis are indexed correctly.

here we added mem.move, so it failed. failing on 153>152 is confusing.

so I just pulled the list of resources - and am asserting on them, and failure will tell which resource is different. 

cc @areshand

File: aptos-move/framework/move-stdlib/tests/mem_tests.move (hunk 2 above)
Comment (igor-aptos) [reply aptos-move/framework/move-stdlib/tests/mem_tests.move:2]: I added some tests, but can add more. 

can you explain when does a &mut become each of the :

* indexer ref - generic
* indexed ref - specialized
* container

**Review by igor-aptos — COMMENTED** [review]
is there a way to unit-test call_native? 

i.e. we are relying for a lot of tests in third_party/move/move-vm/types/src/values/value_tests.rs that are not failing checks internally, to fail checks when native call types get checked. 

Otherwise I am not sure how to have that test coverage, as in end-to-end tests - such code would not compile.

**Comment by igor-aptos** [comment]
addressed all comments and added more tests

File: aptos-move/framework/move-stdlib/sources/mem.move [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+module std::mem {
+    // TODO - functions here are `public(friend)` here for one release,
+    // and to be changed to `public` one release later.
+    #[test_only]
+    friend std::mem_tests;
+
+    /// Swap contents of two passed mutable references.
+    public(friend) native fun swap<T>(left: &mut T, right: &mut T);
+
+    /// Replace value reference points to with the given new value,
```
Comment (georgemitenkov) [inline aptos-move/framework/move-stdlib/sources/mem.move:12]: I think articles are missing, but I am not a native speaker 😄

File: aptos-move/framework/move-stdlib/src/natives/mem.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        .get_feature_flags()
+        .is_native_memory_operations_enabled()
+    {
+        return Err(get_feature_not_available_error());
+    }
+
+    debug_assert!(args.len() == 2);
+
+    if args.len() != 2 {
+        return Err(SafeNativeError::InvariantViolation(PartialVMError::new(
```
Comment (georgemitenkov) [inline aptos-move/framework/move-stdlib/src/natives/mem.rs:56]: Add a message?

File: aptos-move/framework/move-stdlib/tests/mem_tests.move [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    fun test_replace_ints() {
+        let a = 1;
+        let b = 2;
+
+        assert!(replace(&mut a, b) == 1, 0);
+        assert!(a == 2, 1);
+    }
+
+    #[test_only]
+    struct SomeStruct has drop, key {
```
Comment (georgemitenkov) [inline aptos-move/framework/move-stdlib/tests/mem_tests.move:35]: Docstring in `mem.move` says we do this for non-copyable/non-droppable types, but here there is drop?

**Review by igor-aptos — COMMENTED** [review]
oh I missed your previous review, addressing that one and the new one

File: types/src/on_chain_config/aptos_features.rs [hunk 7]
```diff
@@ -95,6 +95,8 @@ pub enum FeatureFlag {
     FEDERATED_KEYLESS = 77,
     TRANSACTION_SIMULATION_ENHANCEMENT = 78,
     COLLECTION_OWNER = 79,
+    /// covers mem::swap and vector::move_range
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply types/src/on_chain_config/aptos_features.rs:100]: Sounds good

File: types/src/on_chain_config/aptos_features.rs [hunk 8]
```diff
@@ -317,6 +320,10 @@ impl Features {
         self.is_enabled(FeatureFlag::TRANSACTION_SIMULATION_ENHANCEMENT)
     }
 
+    pub fn is_native_memory_operations_enabled(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply types/src/on_chain_config/aptos_features.rs:330]: It just reads weird to me because it is plural, but maybe a personal taste

File: aptos-move/aptos-release-builder/src/components/feature_flags.rs [hunk 9]
```diff
@@ -131,6 +131,7 @@ pub enum FeatureFlag {
     FederatedKeyless,
     TransactionSimulationEnhancement,
     CollectionOwner,
+    NativeMemoryOperations,
```
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply aptos-move/aptos-release-builder/src/components/feature_flags.rs:135]: Thanks!
